xdcr.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2024-Present Couchbase, Inc.
+//
+// Use of this software is governed by the Business Source License included
+// in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+// in that file, in accordance with the Business Source License, use of this
+// software will be governed by the Apache License, Version 2.0, included in
+// the file licenses/APL2.txt.
+
+package rosmar
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync/atomic"
+
+	sgbucket "github.com/couchbase/sg-bucket"
+)
+
+// XDCR implements a XDCR bucket to bucket setup within rosmar.
+type XDCR struct {
+	terminator              chan bool
+	fromBucketCollectionIDs map[uint32]sgbucket.DataStoreName
+	fromBucket              *Bucket
+	toBucket                *Bucket
+	replicationID           string
+	docsWritten             atomic.Uint64
+	docsFiltered            atomic.Uint64
+	errorCount              atomic.Uint64
+}
+
+// NewXDCR creates an instance of XDCR backed by rosmar. This is not started until Start is called.
+func NewXDCR(_ context.Context, fromBucket, toBucket *Bucket, opts sgbucket.XDCROptions) (*XDCR, error) {
+	if opts.Mobile != sgbucket.XDCRMobileOn {
+		return nil, errors.New("Only sgbucket.XDCRMobileOn is supported in rosmar")
+	}
+	return &XDCR{
+		fromBucket:              fromBucket,
+		toBucket:                toBucket,
+		replicationID:           fmt.Sprintf("%s-%s", fromBucket.GetName(), toBucket.GetName()),
+		fromBucketCollectionIDs: map[uint32]sgbucket.DataStoreName{},
```

Comment:
What's the expected handling if there are collections defined in fromBucket that aren't present in toBucket?  Should we be checking for this now instead of logging a warning for every mutation that arrives for a non-existent collection?

Reply:
Oh, I see below that we're only supporting the default collection at this point?  (we're not passing any scopes args to the DCP feed).  Are you thinking of adding that later?

Reply:
This was an oversight and I've added tests for the non default collection.